apps/engine/lib/engine/code_intelligence/heex_normalizer.ex

```diff
@@ -0,0 +1,172 @@
+defmodule Engine.CodeIntelligence.HeexNormalizer do
+  @moduledoc false
+
+  alias Forge.Ast
+  alias Forge.Ast.Analysis
+  alias Forge.Document
+  alias Forge.Document.Position
+  alias Forge.Document.Range
+  alias Sourceror.Zipper
+
+  # Matches both opening and closing shorthand components (used for cursor detection)
+  @component_regex ~r/<\/?\.([a-zA-Z0-9_!?.]+)/
+  # Separate regexes for AST normalization to avoid overlap issues
+  @opening_component_regex ~r/<\.([a-zA-Z0-9_!?.]+)/
+  @closing_component_regex ~r/<\/\.([a-zA-Z0-9_!?.]+)/
+  @opening_replacement "< \\1(assigns)"
+  @closing_replacement "</ \\1(assigns)"
+
+  # Normalizes HEEx templates by converting anonymous component references
+  # (e.g., `<.component`) to explicit function calls (e.g., `<component(assigns)`).
+  # It's done in both the AST and document text.
+  #
+  # This allows ElixirSense to understand the shorthand HEEX notation as a local function
+  # (be it imported or not) and return correct location for go-to-definition and hover.
+  #
+  # This normalization is only performed when Phoenix.Component is available in the project
+  # (i.e., phoenix_live_view is in the dependencies).
+  @spec maybe_normalize(Analysis.t(), Position.t()) :: Analysis.t()
+  def maybe_normalize(analysis, position) do
+    if phoenix_component_available?() do
+      new_ast = normalize_ast(analysis, position)
+      new_document = normalize_document(analysis, position)
+      %{analysis | ast: new_ast, document: new_document}
+    else
+      analysis
+    end
+  end
+
+  defp phoenix_component_available? do
+    Code.ensure_loaded?(Phoenix.Component)
```

Comment:
We probably want `Engine.Module.Loader.ensure_loaded?` here. `Code.ensure_loaded?/1` can be quite slow, so we have a cache for it
```suggestion
    Engine.Module.Loader.ensure_loaded?(Phoenix.Component)
```